decred/decred/dcr/blockchain.py

```diff
@@ -0,0 +1,276 @@
+
+        startHeader = self.headerDB[startHash]
+        start = startHeader.height
+        tipHeight = mainchain.last()[0]
+        toScan = tipHeight - start + 1
+        log.info(f"synchronizing {toScan} blocks")
+        rescanBlocks = []
+        while start <= tipHeight:
+            blockHashes = [x[1] for x in mainchain[start : start + maxBlocksPerRescan]]
+            rescanBlocks.extend(self.rpc.rescan(blockHashes))
+            start += maxBlocksPerRescan
+
+        return rescanBlocks
+
+    def syncHeaders(self, root):
+        """
+        Synchronize headers after specified blockHash. It is assumed that root
+        block is a mainchain block, a condition easily met if wallets choose
+        their initial root block to be a few blocks before the tip at wallet
+        creation.
+
+        Args:
+            root (ByteArray): The block hash of the root.
+
+        Returns:
+            list(ByteArray): A list of hashes for blocks removed from mainchain.
+        """
+        synced = 0
+        mainchain = self.mainchainDB
+
+        # Get the root header
+        rootHeader = self.header(root)
+        startTime = time.time()
+
+        # Get the current db tip. Store the root block if mainchain bucket is
+        # empty.
+        try:
+            tipHeight, tipHash = mainchain.last()
+
+        except database.NoValueError:
+
+            # Seed the root block.
+            tipHash = rootHeader.cachedHash()
+            tipHeight = rootHeader.height
+            mainchain[tipHeight] = tipHash
+            log.info(f"stored root block at height {tipHeight}")
+            synced = 1
+
+        # Check for missing blocks before the current root.
+        currentRootHeight, currentRootHash = mainchain.first()
+        if currentRootHeight > rootHeader.height:
+            log.info(f"storing a new root block at height {rootHeader.height}")
+            mainchain[rootHeader.height] = rootHeader.cachedHash()
+            synced += 1 + self.syncHeaderRange(rootHeader.height, currentRootHeight - 1)
+
+        # Prune orphans.
+        orphans = []
+        while self.rpc.getBlockHash(tipHeight) != tipHash:
+            orphans.append(tipHash)
+            del mainchain[tipHeight]
+            tipHeight -= 1
+            tipHash = mainchain[tipHeight]
+
+        # Get the current node tip and synchronize the rest.
+        nodeTip = self.rpc.getBestBlock().height
+        if nodeTip > tipHeight:
+            synced += self.syncHeaderRange(tipHeight, nodeTip)
+
+        secs = int(time.time() - startTime)
+        log.info(f"{secs} seconds to sync {synced} block headers")
+
+        height, blockHash = mainchain.last()
+        log.info(f"best block {reversed(blockHash).hex()} at height {nodeTip}")
+
+        return synced, orphans
+
+    def syncHeaderRange(self, start, end):
+        """
+        Retrieve and store mainchain block headers in the range [start:end),
+        e.g. start is not included.
```

Comment:
I think that (start:end] is the correct notation.
https://en.wikipedia.org/wiki/Bracket_(mathematics)#Intervals